### PR TITLE
feat: try to use ghostty-web for rendering log lines instead of ansiToHtml

### PR DESCRIPTION
**Pull request by abelanger5**
# Description

Switches log line renderer to properly render ANSI control characters using `ghostty-web` instead of ansiToHtml. 

Still to do:
- [x] Light mode
- [x] Fallback when `init` fails (due to no WASM support) - we can just dump to a div

## Type of change

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: frontend/app/src/components/v1/cloud/logging/logs.tsx
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/logs.tsx:1]: think we can remove this file but @mrkaye97 want to double check?

File: frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    () => ({
+      autoScroll,
+      callbacks,
+      theme,
+      onInitError: () => setInitFailed(true),
+    }),
+    [autoScroll, callbacks, theme],
+  );
+
+  useTerminal(containerRef, logs, options);
```
Comment (mrkaye97) [inline frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: are we calling this for side effects?

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
+// @ts-ignore - ghostty-web doesn't have TypeScript declarations
```
Comment (mrkaye97) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts]: 🤮

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  options?: {
+    autoScroll?: boolean;
+    callbacks?: TerminalScrollCallbacks;
+    theme?: TerminalTheme;
+    onInitError?: () => void;
+  },
+) {
+  const [terminalInitialized, setTerminalInitialized] = useState(false);
+  const terminalRef = useRef<any>(null);
+  const fitAddonRef = useRef<any>(null);
```
Comment (mrkaye97) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts]: would be great to not need to use `any` here if possible

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+  }
+  wasmInitPromise = init().then(() => {
+    wasmInitialized = true;
+  });
+  return wasmInitPromise;
+}
+
+export function useTerminal(
+  containerRef: RefObject<HTMLDivElement>,
+  logs: string,
```
Comment (mrkaye97) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:34]: why do we pass this as a string vs a `string[]`?

File: frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+function Terminal({ logs, autoScroll, callbacks, className }: TerminalProps) {
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [initFailed, setInitFailed] = useState(false);
+  const { theme: themeMode } = useTheme();
+
+  // Determine if dark mode
+  const isDark =
+    themeMode === 'dark' ||
```
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx]: should probably just leave the casing as this

File: frontend/app/src/components/v1/cloud/logging/log-viewer.tsx [hunk 6]
```diff
@@ -32,82 +80,46 @@ export function LogViewer({
...
+        ? logs
+        : [
+            {
+              line: emptyMessage,
+              timestamp: new Date().toISOString(),
+              instance: 'Hatchet',
+            },
+          ];
+
+    // Sort by timestamp
```
Comment (mrkaye97) [inline frontend/app/src/components/v1/cloud/logging/log-viewer.tsx]: I dislike comments like this :P

File: frontend/app/src/components/v1/cloud/logging/logs.tsx
Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/logs.tsx:1]: fine with me if it's fine with the type checker

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    return;
+  }
+  if (wasmInitPromise) {
+    return wasmInitPromise;
+  }
+  wasmInitPromise = init().then(() => {
+    wasmInitialized = true;
+  });
+  return wasmInitPromise;
+}
```
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:30]: In hindsight I don't think that this is necessary, init might be safe to call multiple times. The more important thing is that we can only mount a single terminal instance at a time because there's some kind of shared memory magic going on with this wasm init.

**Review by mrkaye97 — COMMENTED** [review]
left a couple comments but broadly looks fine, might be a pain to merge these together 😬

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+export function useTerminal(
+  containerRef: RefObject<HTMLDivElement>,
+  logs: string,
```
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:34]: I would like to change this to an array of string, but I'm keeping it aligned with the previous impl for now

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 9]
```diff
@@ -0,0 +1,302 @@
+import { TerminalTheme } from '../terminalThemes';
+import { init, Terminal, FitAddon } from 'ghostty-web';
+import { useEffect, useRef, RefObject, useState } from 'react';
+
+export interface TerminalScrollCallbacks {
+  onTopReached?: () => void;
+  onBottomReached?: () => void;
```
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:7]: I guess the idea is we can remove these at some point? @mrkaye97

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+        callbacksRef.current?.onInfiniteScroll &&
+        now - lastInfiniteScrollCallRef.current >= 100
+      ) {
+        callbacksRef.current.onInfiniteScroll({
+          scrollTop,
+          scrollHeight,
+          clientHeight,
+        });
+        lastInfiniteScrollCallRef.current = now;
+      }
```
Comment (abelanger5) [inline frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:102]: Need to double check this: I did a bunch of testing on onBottomReached and onTopReached but none on `onInfiniteScroll`.

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 9 above)
Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:7]: I don't think so? because this is for loading more rows as you scroll up or down

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 2 above)
Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts]: this is pure hallucination, it does have TS declarations. I removed this line

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts]: lmao, jesus 😂 

thanks!

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 7 above)
Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:30]: if we really can only call it once, I think it probably should go in a provider at the top level

File: frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx (hunk 1 above)
Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: you mean why is it a hook instead of just building a mega component? we used to have some callbacks returned but no longer.

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 3 above)
Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts]: yeah will take a look

File: frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx (hunk 1 above)
Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: no I mean why are we calling this if it returns nothing 😅

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: if we do need to call it for side effects, there must be a more correct way than this - like maybe inside of a `useMemo` or `useEffect` or something so it only is run on mount

Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: well, the side effect is that it mounts the terminal on the parent component, but a child component doesn't make sense because it won't return anything. but it's complex enough that it needs multiple `useEffect` so a single `useMemo` or `useEffect` doesn't make sense to me - which leaves calling a hook that returns nothing as the last option?

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/components/Terminal.tsx:34]: argh yeah, guess you're right - bummer. maybe just leave a comment for this?

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 9 above)
Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:7]: I thought that was the point of `onInfiniteScroll`? I thought that the caller would be able to case on loading more rows, but it sounds like I'm misunderstanding the purpose of that prop

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:7]: so admittedly I didn't implement this 😅  but I'm pretty sure that onInfiniteScroll might not fire if you're just sitting at the bottom of the scroll view or something - like in the case where you've scrolled all the way down, and then we want to try to fetch some more logs

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:7]: it also could very well not be needed though

File: frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts (hunk 7 above)
Comment (abelanger5) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:30]: I think I'm going to leave it, because I'd like to avoid polluting other parts of the code before I'm set on this approach - but it's a good idea for down the line

Comment (mrkaye97) [reply frontend/app/src/components/v1/cloud/logging/hooks/useTerminal.ts:30]: sounds good 👍
